h2o-core/src/main/java/water/KeyGen.java

```diff
@@ -0,0 +1,115 @@
+package water;
+
+import water.util.ArrayUtils;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public abstract class KeyGen extends Iced<KeyGen> {
+
+  public abstract <T extends Keyed<T>> Key<T> make(Object... params);
+  
+  public static final class ConstantKeyGen extends KeyGen {
+    private final Key key;
+
+    public ConstantKeyGen(Key key) {
+      this.key = key;
+    }
+
+    @Override
+    public <T extends Keyed<T>> Key<T> make(Object... params) {
+      return key;
+    }
+  }
+  
+  public static final class RandomKeyGen extends KeyGen {
+    public RandomKeyGen() {}
+
+    @Override
+    public <T extends Keyed<T>> Key<T> make(Object... params) {
+      return Key.make(Key.rand());
+    }
+  }
+  
+  public static final class PatternKeyGen extends KeyGen {
+    
+    private static final String PIPE = "\\s*\\|\\s*";
+    
+    
+    private enum Command {
+      SUBSTITUTE() {
+        private final Pattern CMD = Pattern.compile("s/(.*?)/(.*?)/?");
```

Comment:
correct, the idea behind those `KeyGen` is to better understand how all those keys are created, and ensure consistency when some keys are created based on other ones. Pipeline can use a lot of keys for all temp frames, for models used in data transformers and so on.

and example of this substitution pattern is used in the pipeline integration in AutoML:
```java
     // in ModelingStep.applyPipeline(...)
      pparams._estimatorKeyGen = hyperParams == null 
              ? new ConstantKeyGen(resultKey) 
              : new PatternKeyGen("{0}|s/"+PIPELINE_KEY_PREFIX+"//")  // in case of grid, remove the Pipeline prefix to obtain the estimator key, this allows naming compatibility with the classic mode.
              ;
```
this way the key formatting doesn't have to leak later in the code that should be pipeline-agnostic